src/neml2/dispatchers/SimpleMPIScheduler.cxx

```diff
@@ -0,0 +1,166 @@
+
+  options.set<std::vector<std::size_t>>("batch_sizes");
+  options.set("batch_sizes").doc() = "List of batch sizes for each device";
+
+  options.set<std::vector<std::size_t>>("capacities") = {};
+  options.set("capacities").doc() = "List of capacities for each device, default to batch_sizes";
+
+  return options;
+}
+
+SimpleMPIScheduler::SimpleMPIScheduler(const OptionSet & options)
+  : WorkScheduler(options),
+    _available_devices(options.get<std::vector<Device>>("devices")),
+    _batch_sizes(options.get<std::vector<std::size_t>>("batch_sizes")),
+    _capacities(options.get("capacities").user_specified()
+                    ? options.get<std::vector<std::size_t>>("capacities")
+                    : _batch_sizes),
+    _comm(TIMPI::Communicator(MPI_COMM_WORLD))
+{
+  neml_assert(_available_devices.size() == _batch_sizes.size(),
+              "Number of batch sizes must match the number of devices.");
+  neml_assert(
+      _available_devices.size() == _capacities.size(),
+      "Number of capacities must match the number of devices and the number of batch sizes.");
+}
+
+void
+SimpleMPIScheduler::setup()
+{
+  WorkScheduler::setup();
+
+  // First pass:
+  // - Prohibit any CPU
+  // - Check if any CUDA device is present
+  bool cuda = false;
+  for (const auto & device : _available_devices)
+  {
+    neml_assert(!device.is_cpu(), "CPU device is not allowed in SimpleMPIScheduler");
+    if (device.is_cuda())
+      cuda = true;
+    else
+      neml_assert(false, "Unsupported device type: ", device);
+  }
+
+  // Second pass:
+  // - If multiple CUDA devices are present, make sure each CUDA device has a concrete
+  //   (nonnegative), unique device ID
+  bool has_multiple_cuda_devices = _available_devices.size() > 1 && cuda;
+  if (has_multiple_cuda_devices)
+  {
+    std::set<DeviceIndex> cuda_device_ids;
+    for (const auto & device : _available_devices)
+    {
+      auto device_id = device.index();
+      neml_assert(device_id >= 0, "Device ID must be nonnegative");
+      neml_assert(cuda_device_ids.find(device_id) == cuda_device_ids.end(),
+                  "Device ID must be unique. Found duplicate: ",
+                  device_id);
+      cuda_device_ids.insert(device_id);
+    }
+  }
+
+  determine_my_device();
+}
+
+void
+SimpleMPIScheduler::determine_my_device()
+{
+  // Get the hostname via hwloc and hash it into an MPI_INT
+  hwloc_topology_t topology;
+  hwloc_obj_t machine;
+
+  hwloc_topology_init(&topology); // initialization
+  hwloc_topology_load(topology);  // actual detection
+  machine = hwloc_get_root_obj(topology);
+  std::hash<std::string> hasher;
+  std::string hostname(hwloc_obj_get_info_by_name(machine, "HostName"));
+  int id = static_cast<int>(hasher(hostname) % std::numeric_limits<int>::max());
+  hwloc_topology_destroy(topology);
+
+  // Make a new communicator based on this hashed hostname
+  TIMPI::Communicator new_comm;
+  _comm.split(id, _comm.rank(), new_comm);
```

Comment:
Seems like the splitting is based purely on `id` which is hashed from hostname. I think `MPI_Get_processor_name` can do the same job and is portable, at least for both MPICH and OpenMPI.

Reply:
Yep